pkg/operator/operator.go

```diff
@@ -0,0 +1,120 @@
+	// objects. It is split in the dimensions of Kubernetes objects and
+	// corresponding actions (add, delete, update).
+	triggerByCounter *prometheus.CounterVec
+}
+
+func NewMetrics(name string, r prometheus.Registerer) *Metrics {
+	reg := prometheus.WrapRegistererWith(prometheus.Labels{"controller": name}, r)
+	m := Metrics{
+		reg: reg,
+		reconcileErrorsCounter: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_operator_reconcile_errors_total",
+			Help: "Number of errors that occurred while reconciling the statefulset",
+		}),
+		triggerByCounter: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "prometheus_operator_triggered_total",
+			Help: "Number of times a Kubernetes object add, delete or update event" +
+				" triggered the Prometheus Operator to reconcile an object",
+		}, []string{"triggered_by", "action"}),
+		stsDeleteCreateCounter: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_operator_reconcile_sts_delete_create_total",
+			Help: "Number of times that reconciling a statefulset required deleting and re-creating it",
+		}),
+		listCounter: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_operator_list_operations_total",
+			Help: "Total number of list operations",
+		}),
+		listFailedCounter: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_operator_list_operations_failed_total",
+			Help: "Total number of list operations that failed",
+		}),
+	}
+	m.reg.MustRegister(
+		m.reconcileErrorsCounter,
+		m.triggerByCounter,
+		m.stsDeleteCreateCounter,
+		m.listCounter,
+		m.listFailedCounter,
+	)
+	return &m
+}
+
+func (m *Metrics) ReconcileErrorsCounter() prometheus.Counter {
+	return m.reconcileErrorsCounter
+}
+
+func (m *Metrics) StsDeleteCreateCounter() prometheus.Counter {
+	return m.stsDeleteCreateCounter
+}
+
+func (m *Metrics) TriggerByCounter(triggered_by, action string) prometheus.Counter {
+	return m.triggerByCounter.WithLabelValues(triggered_by, action)
+}
+
+func (m *Metrics) MustRegister(metrics ...prometheus.Collector) {
+	m.reg.MustRegister(metrics...)
+}
+
+type instrumentedListerWatcher struct {
+	next   cache.ListerWatcher
+	total  prometheus.Counter
+	failed prometheus.Counter
+}
+
+// NewInstrumentedListerWatcher returns a cache.ListerWatcher with instrumentation.
+func (m *Metrics) NewInstrumentedListerWatcher(lw cache.ListerWatcher) cache.ListerWatcher {
+	return &instrumentedListerWatcher{
+		next:   lw,
+		total:  m.listCounter,
+		failed: m.listFailedCounter,
+	}
+}
+
+// List implements the cache.ListerWatcher interface.
+func (i *instrumentedListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	i.total.Inc()
+	ret, err := i.next.List(options)
+	if err != nil {
+		i.failed.Inc()
+	}
+	return ret, err
+}
+
+// Watch implements the cache.ListerWatcher interface.
+func (i *instrumentedListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	return i.next.Watch(options)
```

Comment:
It would make sense to also add the same metrics for watch, not necessarily to do it in this PR, we can also open an issue for it. WDYT?